nameko/extensions.py

```diff
@@ -0,0 +1,308 @@
+"""
+Provides classes and method to deal with dependency injection.
+"""
+from __future__ import absolute_import
+
+from functools import partial
+import inspect
+import types
+import weakref
+
+from eventlet.event import Event
+
+from logging import getLogger
+_log = getLogger(__name__)
+
+
+ENTRYPOINT_EXTENSIONS_ATTR = 'nameko_entrypoints'
+
+shared_extensions = weakref.WeakKeyDictionary()
+
+
+class Extension(object):
+
+    __clone = False
+    __params = None
+
+    def __new__(cls, *args, **kwargs):
+        inst = super(Extension, cls).__new__(cls, *args, **kwargs)
+        inst.__params = (args, kwargs)
+        return inst
+
+    def __init__(self, *args, **kwargs):
+        """ Note that Extension.__init__ is called during :meth:`clone` as
+        well as at instantiation time, so avoid side-effects in this method.
+        Use :meth:`setup` instead.
+        """
+        super(Extension, self).__init__(*args, **kwargs)
+
+    def setup(self, container):
+        """ Called before the service container starts.
+
+        Extensions should do any required initialisation here.
+        """
+
+    def start(self):
+        """ Called when the service container has successfully started.
+
+        This is only called after all other Extensions have successfully
+        returned from :meth:`Extension.setup`. If the Extension reacts
+        to external events, it should now start acting upon them.
+        """
+
+    def stop(self):
+        """ Called when the service container begins to shut down.
+
+        Extensions should do any graceful shutdown here.
+        """
+
+    def kill(self):
+        """ Called to stop this extension without grace.
+
+        Extensions should urgently shut down here. This means
+        stopping as soon as possible by omitting cleanup.
+        This may be distinct from ``stop()`` for certain dependencies.
+
+        For example, :class:`~messaging.QueueConsumer` tracks messages being
+        processed and pending message acks. Its ``kill`` implementation
+        discards these and disconnects from rabbit as soon as possible.
+
+        Extensions should not raise during kill, since the container
+        is already dying. Instead they should log what is appropriate and
+        swallow the exception to allow the container kill to continue.
+        """
+
+    def clone(self, container):
+        if self.is_clone:
+            raise RuntimeError('Cloned extensions cannot be cloned.')
```

Comment:
do we have some common errors that might cause this worth mentioning? (a la "did you ..."?)